qiskit_experiments/warnings.py

```diff
@@ -0,0 +1,245 @@
+        use_instead: Alternative approach.
+
+    Examples:
+
+        .. code-block::
+
+            @deprecated_logic(version_removed="0.3", use_instead="use new_function")
+            def old_function(*args, **kwargs):
+                pass
+
+            def new_function(*args, **kwargs):
+                pass
+
+    Returns:
+        Deprecated function or method.
+    """
+
+    def deprecated_wrapper(func: Callable):
+        @functools.wraps(func)
+        def _wrap(*args, **kwargs):
+            namespace = func.__qualname__.split(".")
+            if len(namespace) == 1:
+                message = f"The function '{func.__name__}' has been deprecated and "
+            else:
+                cls_name, meth_name = namespace
+                message = f"The method '{meth_name}' of '{cls_name}' class has been deprecated and "
+            if version_removed:
+                message += f"will be removed in Qiskit Experiments {version_removed}. "
+            else:
+                message += "will be removed in future release. "
+            if use_instead:
+                message += f"Please '{use_instead}' instead. "
+            warnings.warn(message, DeprecationWarning, stacklevel=2)
+            return func(*args, **kwargs)
+
+        return _wrap
+
+    return deprecated_wrapper
+
+
+def deprecated_class(
+    version_removed: Optional[str] = None,
+    new_cls: Optional[Type] = None,
+) -> Callable:
+    """A class decorator to show deprecation warning and
+    patch __new__ method of the class to instantiate the new class.
+
+    Args:
+        version_removed: The Qiskit Experiment version that this class is removed.
+        new_cls: Alternative class type.
+
+    Examples:
+
+        .. code-block::
+
+            @deprecated_class(version_removed="0.3", new_cls=NewCls)
+            class OldClass:
+                pass
+
+            class NewClass:
+                pass
+
+    Returns:
+        Deprecated class.
+    """
+
+    def patch_new(cls) -> Type:
+        @functools.wraps(cls.__init__, assigned=("__annotations__",))
+        def new(deprecated_cls, *args, **kwargs):
+            message = f"Class '{deprecated_cls.__name__}' has been deprecated"
+            if new_cls:
+                message += f" and replaced with '{new_cls.__name__}'. "
+            else:
+                message += ". "
+            if version_removed:
+                message += f"This class will be removed in Qiskit Experiments {version_removed}. "
+            else:
+                message += "This class will be removed in future release. "
+            warnings.warn(message, DeprecationWarning, stacklevel=2)
+            instance = object.__new__(new_cls or deprecated_cls)
+            instance.__init__(*args, **kwargs)
+            return instance
+
+        cls.__new__ = new
+        return cls
+
+    return patch_new
+
+
+def deprecated_options(
+    options_map: Dict[str, str],
+    version_removed: Optional[str] = None,
```

Comment:
This should also allow a custom message so we can deprecate options that aren't a simple renaming (ie the new option might take a different value type or there isnt a 1-to-1 replacement)